### ISSUE TITLE
spec issue: maximum message sizing

### ISSUE THREAD

**Issue by jlevon** [nutanix/libvfio-user#469]
From Stefan:

```
> > > How do potentially large messages work around max_msg_size? It is hard
> > > for the client/server to anticipate the maximum message size that will
> > > be required ahead of time, so they can't really know if they will hit a
> > > situation where max_msg_size is too low.
> > 
> > Are there specific messages you're worried about? would it help to add a
> > specification stipulation as to minimum size that clients and servers must
> > support?
> > 
> > Ultimately the max msg size exists solely to ease implementation: with a
> > reasonable fixed size, we can always consume the entire data in one go, rather
> > than doing partial reads. Obviously that needs a limit to avoid unbounded
> > allocations.
> 
> It came to mind when reading about the dirty bitmap messages. Memory
> dirty bitmaps can become large. An 8 GB memory region has a 1 MB dirty
> bitmap.
```

**Comment by tmakatos** [comment]
The original reason this limit was introduced was to avoid dynamically allocating memory in libvfio-user in order to improve performance.

Discussed this with @jlevon, why should drop this and receive/process the message in a piece-wise manner. The maximum size of the buffer can be configurable by the user. I'll start implementing this in libvfio-user and see how it looks like and if it works we can update the spec.

**Comment by tmakatos** [comment]
Since the message header contains the total message size, can this be as simple as the sender sending the payload in whatever piece-wise manner it wants? Similarly, the server knows the total message size from the header so it can receive the payload in a piece-wise manner. We just need to make sure than the sender doesn't send anything else in between. @jlevon what do you think?

**Comment by jlevon** [comment]
This really applies only to DMA_UNMAP and REGION/DMA read/write today, the others are either fixed enough, have natural limits, or have the argsz prefix.

Trying to do this on the server side is pretty tedious: instead of being able to assemble a reply, we'd have to provide some streaming interface instead when e.g. assembling the dirty bitmaps. (It's much easier to do for REGION_READ and DMA_WRITE certainly).

We also need to have that interface for the other way around: instead of being able to just read the message body, we have to just read the header then push down the payload read into every caller.

But this is much worse on the qemu side: instead of being able to wrap behind the proxy, now we'd have to expose the different behaviour (ioctl in/out versus streaming) all the way up to the caller.

Contrast with an enforced max msg size with a reasonable default: it's much easier to split up an outgoing request into multiple request/replies based upon the min(server_msg_size, client_msg_size).

**Comment by jlevon** [comment]
New plan: add a max-transfer-size that *only* applies to DMA_READ/WRITE and REGION_READ/WRITE. Clients and servers must use one of the client or server values to split up larger accesses.

Everything else should be handled based on the expected reply size, or other limits (such as the 2048 max IRQS, for example).

**Comment by tmakatos** [comment]
@jlevon is this done?

**Comment by jlevon** [comment]
Fixed by #541
